src/dcap_provider.cpp

```diff
@@ -1607,6 +1600,107 @@ bool check_cache(std::string cached_file_name, sgx_ql_config_t** pp_quote_config
+    retval = convert_to_intel_error(err);
+    if (retval == SGX_QL_SUCCESS)
+    {
+        log(SGX_QL_LOG_INFO, "Successfully parsed certificate chain: %s.", cert_data.c_str());
+    }
+    else
+    {
+        log(SGX_QL_LOG_ERROR, "Unable to parse certificate chain from the response.");
+        return retval;
+    }
+
+    // copy the null-terminator for convenience (less error-prone)
+    const uint32_t cert_data_size =
+        static_cast<uint32_t>(cert_data.size()) + 1;
+
+    // allocate return value contiguously (makes caching easier)
+    const size_t buf_size = sizeof(sgx_ql_config_t) + cert_data_size;
+    uint8_t* buf = new uint8_t[buf_size];
+    memset(buf, 0, buf_size);
+
+#ifndef NDEBUG
+    const uint8_t* buf_end = buf + buf_size;
+#endif
+
+    *pp_quote_config = reinterpret_cast<sgx_ql_config_t*>(buf);
+    buf += sizeof(sgx_ql_config_t);
+    assert(buf <= buf_end);
+    (*pp_quote_config)->cert_cpu_svn = temp_config.cert_cpu_svn;
+    (*pp_quote_config)->cert_pce_isv_svn = temp_config.cert_pce_isv_svn;
+    (*pp_quote_config)->version = SGX_QL_CONFIG_VERSION_1;
+    (*pp_quote_config)->p_cert_data = buf;
+    (*pp_quote_config)->cert_data_size = cert_data_size;
+    memcpy(
+        (*pp_quote_config)->p_cert_data, cert_data.data(), cert_data_size);
+    buf += cert_data_size;
+    assert(buf == buf_end);
+
+    // Get the cache control header
+    std::string cache_control;
+    auto get_cache_header_operation = extract_from_json(
+        json_body, headers::CERT_CACHE_CONTROL, &cache_control);
+    retval = convert_to_intel_error(get_cache_header_operation);
+    if (retval == SGX_QL_SUCCESS)
+    {
+        log(SGX_QL_LOG_INFO,
+            "%s : %s",
+            headers::CERT_CACHE_CONTROL,
+            cache_control.c_str());
+        time_t expiry = 0;
+        if (get_cert_cache_expiration_time(cache_control, cached_file_name, expiry))
+        {
+			local_cache_add(cached_file_name, expiry, buf_size, *pp_quote_config);
+			log(SGX_QL_LOG_INFO, "Stored certificate in cache in the following location: %s.", get_cached_file_location(cached_file_name).c_str());
+		}
+		else 
+		{
+			log(SGX_QL_LOG_ERROR, "Unable to retrieve the certificate expiry when writing to local cache.");
+		}
+    }
+    else
+    {
+        log(SGX_QL_LOG_ERROR, "Unable to add certificate to local cache.");
+    }
+
+	retval = SGX_QL_SUCCESS;
+
+	return retval;
+}
+
+extern "C" bool store_certificate(const std::string& qe_id, const std::string& cpu_svn, const std::string& pce_svn, const std::string& pce_id, const std::string& response_body)
+{
+	bool result = false;
+	sgx_ql_config_t* pp_quote_config = nullptr;
+	nlohmann::json json_body = nlohmann::json::parse(response_body);
+
+	std::stringstream cached_file_name = build_cache_url(qe_id, cpu_svn, pce_svn, pce_id);
+
+	if (store_certificate_internal(cached_file_name.str(), json_body, &pp_quote_config) == SGX_QL_SUCCESS) {
+		result = true;
+	}
+
+	delete[] pp_quote_config;

```

Comment:
This free is going to be undefined behaviour because allocation was not done for `new sqx_ql_config_t[]`. 

I am not 100% sure but I think this needs to be:

`delete pp_quote_config`  or `delete[] reinterpret_cast<uint8_t*>(pp_quote_config)` 

You could look at how QGSD frees its memory too as a reference.

Reply:
This is imitating the way memory is freed by sgx_ql_free_quote_config when called by a caller that had previous allocated the memory of pp_quote_config with sgx_ql_get_quote_config, which was already working previously, so I think it should work properly.